File: gbd_2021/shared_code/central_comp/risk/sev/utils/risk_prevalence.R
#' For a given continuous risk and demographic(s), pull risk exposure and exposure SD,
#' generate a distribution of exposure and calculate prevalence above or below some
#' threshold within that distribution.
#' 
#' Implemented for FPG and SBP.
#'
#' @return data.table of prevalence above/below exposure threshold for the given risk
#' and demographics
get_risk_prevalence <- function(
    rei_id, rei, location_id, year_id, gbd_round_id, decomp_step, n_draws
) {
    # load needed libraries, some from PAF code
    source("FILEPATH/get_draws.R")
    source("FILEPATH/data.R")
    source("FILEPATH/db.R")
    source("FILEPATH/edensity.R")
    
    # Confirm REI has implementation in this function
    if (!(rei %in% c("metab_fpg", "metab_sbp"))) {
        stop(paste0("Function not implemented for rei '", rei,"'."))
    }
    
    # pull exposure and exposure SD
    exposure_dt <- merge(
        get_exp(
            rei_id = rei_id, location_id = location_id, year_id = year_id, sex_id = c(1, 2),
            gbd_round_id = gbd_round_id, decomp_step = decomp_step, n_draws = n_draws
        ),
        get_exp_sd(
            rei_id = rei_id, location_id = location_id, year_id = year_id, sex_id = c(1, 2),
            gbd_round_id = gbd_round_id, decomp_step = decomp_step, n_draws = n_draws
        ),
        by = c("location_id", "year_id", "age_group_id", "sex_id", "draw")
    )

    # Add on exposure threshold
    # For FPG: diabetic > 7 mmol/L FPG
    # For SBP: HTN HD exposure level > 140 mmHg
    rei_metadata <- get_rei_meta(rei_id = rei_id, gbd_round_id = gbd_round_id)
    if (rei == "metab_fpg") {
        exposure_dt[, threshold := 7]
    } else if (rei == "metab_sbp") {
        exposure_dt[, threshold := 140]
    }

    # calculate risk prevalence ABOVE or BELOW threshold
    ensemble_weights <- fread(paste0("FILEPATH/_weights/", rei, ".csv"))
    calc_prevalence <- function(i) {
        weights <- ensemble_weights[
            age_group_id == exposure_dt[i, ]$age_group_id & sex_id == exposure_dt[i, ]$sex_id,
        ][, c("location_id", "year_id", "age_group_id", "sex_id") := NULL][1, ]
        exposure_threshold <- exposure_dt[i, ]$threshold
        dens <- get_edensity(weights, exposure_dt[i, ]$exp_mean, exposure_dt[i, ]$exp_sd)
        if (!rei_metadata$inv_exp) {
            # Area above threshold
            return(pracma::trapz(dens$x[dens$x > exposure_threshold], dens$fx[dens$x > exposure_threshold]))
        } else {
            # Area under threshold
            return(pracma::trapz(dens$x[dens$x < exposure_threshold], dens$fx[dens$x < exposure_threshold]))
        }
        
    }
    calc_prevalencec <- compiler::cmpfun(calc_prevalence)
    prevs <- mclapply(1:nrow(exposure_dt), calc_prevalencec, mc.cores = 2)
    prevalence_dt <- cbind(exposure_dt, prevalence = unlist(prevs))
    return(prevalence_dt[, .(location_id, year_id, age_group_id, sex_id, draw, prevalence)])
}

#' Pull iron deficiency prevalence, defined as the sum of moderate and severe anemia prevalence.
#' 
#' If all the years requested are estimation years, pull estimates directly from get_draws.
#' If not, we interpolate from the minimum given year - 5 to the maximum given year + 5
#' (rounded to the closest estimation years).
#' 
#' @return data.table of prevalence of moderate or severe anemia for given risk demographics
#' in rate - space
get_iron_deficiency_prevalance <- function(location_id, year_id, gbd_round_id, decomp_step, n_draws,
                                           df_utils) {
    source("FILEPATH/get_draws.R")
    source("FILEPATH/get_demographics.R")
    source("FILEPATH/interpolate.R")
    
    # Get estimation years: if year_id is one of them, call get_draws. Otherwise, interpolate
    estimation_year_ids <- get_demographics("epi", gbd_round_id = gbd_round_id)$year_id

    # Pull moderate and severe anemia prevalence draws
    mod_anemia_me_id <- 1622
    sev_anemia_me_id <- 1623
    if (all(year_id %in% estimation_year_ids)) {
        anemia <- get_draws("modelable_entity_id", gbd_id = c(mod_anemia_me_id, sev_anemia_me_id),
                            source = "epi", measure_id = 5, metric_id = 3, location_id = location_id,
                            year_id = year_id, gbd_round_id = gbd_round_id, decomp_step = decomp_step,
                            n_draws = n_draws, downsample = n_draws != 1000)
    } else {
        # Interpolate +/- 5 years from min and max of given years, one ME at a time
        # Here we calculate the closest estimation years to min(year_id) - 5 and max(year_id) + 5 
        estimation_years <- data.table(estimation_year_ids = estimation_year_ids, val = estimation_year_ids)
        setattr(estimation_years, "sorted", "estimation_year_ids") 
        reporting_year_start <- estimation_years[J(min(year_id) - 5), roll = "nearest"]$val
        reporting_year_end <- estimation_years[J(max(year_id) + 5), roll = "nearest"]$val
        message(paste("Interpolating anemia prevalence between", reporting_year_start, "and", reporting_year_end))
        
        mod_anemia <- interpolate(
            gbd_id_type = "modelable_entity_id",
            gbd_id = mod_anemia_me_id,
            source = "epi",
            measure_id = 5,
            location_id = location_id,
            gbd_round_id = gbd_round_id,
            decomp_step = decomp_step,
            reporting_year_start = reporting_year_start,
            reporting_year_end = reporting_year_end,
        )
        sev_anemia <- interpolate(
            gbd_id_type = "modelable_entity_id",
            gbd_id = sev_anemia_me_id,
            source = "epi",
            measure_id = 5,
            location_id = location_id,
            gbd_round_id = gbd_round_id,
            decomp_step = decomp_step,
            reporting_year_start = reporting_year_start,
            reporting_year_end = reporting_year_end,
        )
        
        # Subset to only the year requested
        year <- year_id
        anemia <- rbind(mod_anemia[year_id %in% year], sev_anemia[year_id %in% year])
        
        # Downsample draws if necessary using sample method as get_draws
        anemia <- downsample(dt=anemia, n_draws=n_draws, df_utils=df_utils)
    }
    
    # Sum moderate and severe prevalence rates together
    draw_cols <- paste0("draw_", 0:(n_draws - 1))
    anemia <- anemia[, lapply(.SD, sum), by = c("age_group_id", "location_id", "sex_id", "year_id"),
                     .SDcols = draw_cols]
    
    # Reshape long and return
    anemia <- melt(anemia, id.vars = c("age_group_id", "sex_id", "location_id", "year_id"),
                   variable.name = "draw", value.name = "prevalence")
    anemia[, draw := as.integer(gsub("draw_", "", draw))]
    
    return(anemia)
}

File: gbd_2023/shared_code/central_comp/risk/sev/lib/risk_prevalence.py
from typing import Dict, List

import pandas as pd

import chronos
from gbd.constants import measures, metrics
from get_draws import api as get_draws_api

from ihme_cc_sev_calculator.lib import constants

# Map of rei_id -> definitional threshold
THRESHOLD_MAP: Dict[int, int] = {
    constants.HIGH_FPG_REI_ID: 7,  # FPG: diabetic > 7 mmol/L FPG
    constants.HIGH_SBP_REI_ID: 140,  # SBP: hypertensive heart disease > 140 mmHg
}


def get_iron_deficiency_prevalance(
    location_id: int,
    year_ids: List[int],
    estimation_year_ids: List[int],
    release_id: int,
    me_ids: pd.DataFrame,
    n_draws: int,
    draw_cols: List[str],
) -> pd.DataFrame:
    """Get iron deficiency prevalence.

    Defined as the prevalence of moderate or severe anemia.

    If requesting years outside of estimation years, the function interpolates
    between the closest estimation years to min(year_ids) - 5 and max(year_ids) + 5.
    """
    query_str = (
        "modelable_entity_id.isin("
        f"{[constants.MODERATE_ANEMIA_ME_ID, constants.SEVERE_ANEMIA_ME_ID]})"
    )
    prevalence_me_ids = me_ids.query(query_str)
    if len(prevalence_me_ids) != 2:
        raise RuntimeError(f"Expected exactly two anemia MEs. Received:\n{prevalence_me_ids}")

    if set(year_ids).issubset(estimation_year_ids):
        # If all the requested years are estimation years, we can pull existing results
        prevalence = get_draws_api.get_draws(
            "modelable_entity_id",
            gbd_id=prevalence_me_ids["modelable_entity_id"].tolist(),
            version_id=prevalence_me_ids["model_version_id"].tolist(),
            source="epi",
            location_id=location_id,
            year_id=year_ids,
            measure_id=measures.PREVALENCE,
            metric_id=metrics.RATE,
            release_id=release_id,
            n_draws=n_draws,
            downsample=True,
        )
    else:
        # If any of the requested years fall outside of estimation years, we interpolate
        # between the closest estimation years to min(year_ids) - 5 and max(year_ids) + 5
        start_year_id = _get_closest_value(estimation_year_ids, target=min(year_ids) - 5)
        end_year_id = _get_closest_value(estimation_year_ids, target=max(year_ids) + 5)

        moderate_prevalence = chronos.interpolate(
            gbd_id_type="modelable_entity_id",
            gbd_id=constants.MODERATE_ANEMIA_ME_ID,
            version_id=prevalence_me_ids.query(
                f"modelable_entity_id == {constants.MODERATE_ANEMIA_ME_ID}"
            )["model_version_id"].iat[0],
            source="epi",
            location_id=location_id,
            measure_id=measures.PREVALENCE,
            release_id=release_id,
            reporting_year_start=start_year_id,
            reporting_year_end=end_year_id,
            n_draws=n_draws,
            downsample=True,
        )
        severe_prevalence = chronos.interpolate(
            gbd_id_type="modelable_entity_id",
            gbd_id=constants.SEVERE_ANEMIA_ME_ID,
            version_id=prevalence_me_ids.query(
                f"modelable_entity_id == {constants.SEVERE_ANEMIA_ME_ID}"
            )["model_version_id"].iat[0],
            source="epi",
            location_id=location_id,
            measure_id=measures.PREVALENCE,
            release_id=release_id,
            reporting_year_start=start_year_id,
            reporting_year_end=end_year_id,
            n_draws=n_draws,
            downsample=True,
        )

        # Subset to only requested years and downsample
        prevalence = (
            pd.concat([moderate_prevalence, severe_prevalence])
            .query(f"year_id.isin({year_ids})")
            .reset_index(drop=True)
        )

    # Verify all requested years are present
    different_year_ids = set(year_ids).symmetric_difference(prevalence["year_id"])
    if different_year_ids:
        raise RuntimeError(
            "Unexpected symmetric difference in requested years and years found in draws: "
            f"{different_year_ids}."
        )

    # Sum moderate and severe anemia prevalences together (assuming mutual exclusivity)
    return prevalence.groupby(constants.DEMOGRAPHIC_COLS)[draw_cols].sum().reset_index()


def _get_closest_value(values: List[int], target: int) -> int:
    """Get closest value in values to target.

    Ex:
        If values=[0, 5, 10], target=4 => returns 5.
    """
    return min(values, key=lambda value: abs(value - target))


def format_exposure_for_edensity(
    rei_id: int, exposure: pd.DataFrame, exposure_sd: pd.DataFrame, draw_cols: List[str]
) -> pd.DataFrame:
    """Format exposure and exposure SD draws for use within edensity R code.

    Melts the two dataframes long, joins, and adds an exposure threshold column based on
    the REI:
        * FPG: > 7 mmol/L FPG
        * SBP: > 140 mmHg

    Returns:
        DataFrame with columns location_id, year_id, age_group_id, sex_id, draw, exp_mean,
            exp_sd, threshold
    """
    exposure_long = exposure.melt(
        id_vars=constants.DEMOGRAPHIC_COLS,
        value_vars=draw_cols,
        var_name="draw",
        value_name="exp_mean",
    )
    exposure_sd_long = exposure_sd.melt(
        id_vars=constants.DEMOGRAPHIC_COLS,
        value_vars=draw_cols,
        var_name="draw",
        value_name="exp_sd",
    )

    exposure_df = pd.merge(
        exposure_long, exposure_sd_long, on=constants.DEMOGRAPHIC_COLS + ["draw"]
    )

    # Add on exposure threshold. FPG and SBP outcomes are definitional, meaning that they
    # are defined as exposure above a threshold.
    exposure_df = exposure_df.assign(threshold=THRESHOLD_MAP[rei_id])

    return exposure_df
